company.el

```diff
@@ -655,19 +655,109 @@ commands in the `company-' namespace, abort completion."
                         (repeat :tag "Commands" function))
                  (repeat :tag "Commands" function)))
 
-(defcustom company-show-numbers nil
-  "If enabled, show quick-access numbers for the first ten candidates."
-  :type '(choice (const :tag "off" nil)
-                 (const :tag "left" 'left)
-                 (const :tag "on" 't)))
+(defun company-custom--set-quick-access (option value)
+  "Re-bind quick-access key sequences on OPTION VALUE change."
+  (when (boundp 'company-active-map)
+    (company-keymap--unbind-quick-access company-active-map))
+  (when (boundp 'company-search-map)
+    (company-keymap--unbind-quick-access company-search-map))
+  (custom-set-default option value)
+  (when (boundp 'company-active-map)
+    (company-keymap--bind-quick-access company-active-map))
+  (when (boundp 'company-search-map)
+    (company-keymap--bind-quick-access company-search-map)))
+
+(defcustom company-quick-access-keys '("1" "2" "3" "4" "5" "6" "7" "8" "9" "0")
+  "Character strings used as a part of quick-access key sequences.
+To change this value without Customize interface, use `customize-set-variable'.
+
+To change a quick-access key sequences modifier, customize
+`company-quick-access-modifier'.
+
+If `company-show-quick-access' is enabled, show quick-access hints
+beside the candidates."
+  :set #'company-custom--set-quick-access
+  :type '(choice
+          (const :tag "Digits" ("1" "2" "3" "4" "5" "6" "7" "8" "9" "0"))
+          (const :tag "QWERTY home row" ("a" "s" "d" "f" "g" "h" "j" "k" "l" ";"))
+          (const :tag "Dvorak home row" ("a" "o" "e" "u" "i" "d" "h" "t" "n" "s"))
+          (repeat :tag "User defined" string))
+  :package-version '(company . "0.9.14"))
+
+(defcustom company-quick-access-modifier "M"
```

Comment:
Not what we discussed, but works for me. :ok_hand:

Reply:
Do you mean [symbols usage](https://github.com/company-mode/company-mode/pull/1115#issuecomment-868698762)? Unfortunately, I couldn't make them work, so just used the approach that worked. (yeah, my ELisp-fu isn't strong enough yet.)

As for the usage of characters for the keys setup (instead of strings), after thinking more about it, I concluded that they seemed to be potentially more confusing (especially to newcomers). 

And as for the discussed earlier quick-access keys validation, it worked fine by default when assigning keys from Customize: the bindings are re-added and if something goes wrong the error is shown in the minibuffer. That seemed to be enough to me.

That's all I could think of regarding "Not what we discussed". Let me know if that's something else to avoid any miscommunication.

Reply:
> Do you mean symbols usage? Unfortunately, I couldn't make them work, so just used the approach that worked. (yeah, my ELisp-fu isn't strong enough yet.)

That's the one, yes.

It's not a big problem, but in general I like to follow existing examples (e.g. `cua-rectangle-modifier-key`) in such situations, if only to reduce arbitrary differences between packages.

The code examples out there construct binding vectors with it, which I agree is non-trivial and requires knowledge of the respective data structures (which the use of `kbd` helpfully avoids). But we can use a simple alist to translate from one set of values to another (100% untested code):

```el
(defun company-keymap--bind-quick-access (keymap)
  (let ((modifier-char (company-keymap--quick-access-modifier)))
    (dolist (key company-quick-access-keys)
      (define-key keymap
        (company-keymap--kbd-quick-access modifier-char key)
        #'company-complete-quick-access))))

(defun company-keymap--quick-access-modifier ()
  (let ((modifier-char (assoc-default company-quick-access-modifier
                                      '((meta . "M")
                                        (control . "C")
                                        (shift . "S")
                                        (hyper . "h")
                                        (super . "s")))))
    (warn "company-quick-access-modifier value unknown: %S" company-quick-access-modifier)
    modifier-char))

(defun company-keymap--unbind-quick-access (keymap)
  (let ((modifier-char (company-keymap--quick-access-modifier)))
    (dolist (key company-quick-access-keys)
      (define-key keymap (company-keymap--kbd-quick-access modifier-char key) nil))))

(defun company-keymap--kbd-quick-access (key)
  (kbd (format "%s-%s" company-quick-access-modifier key)))
```